server/src/main/java/org/elasticsearch/transport/TcpTransportKeepAlive.java

```diff
@@ -0,0 +1,204 @@
+
+    private final Logger logger = LogManager.getLogger(TcpTransportKeepAlive.class);
+    private final CounterMetric successfulPings = new CounterMetric();
+    private final CounterMetric failedPings = new CounterMetric();
+    private final ConcurrentMap<TimeValue, ScheduledPing> pingIntervals = ConcurrentCollections.newConcurrentMap();
+    private final Lifecycle lifecycle = new Lifecycle();
+    private final ThreadPool threadPool;
+    private final PingSender pingSender;
+    private final BytesReference pingMessage;
+
+    TcpTransportKeepAlive(ThreadPool threadPool, PingSender pingSender) {
+        this.threadPool = threadPool;
+        this.pingSender = pingSender;
+
+        try (BytesStreamOutput out = new BytesStreamOutput()) {
+            out.writeByte((byte) 'E');
+            out.writeByte((byte) 'S');
+            out.writeInt(PING_DATA_SIZE);
+            pingMessage = out.bytes();
+        } catch (IOException e) {
+            throw new AssertionError(e.getMessage(), e); // won't happen
+        }
+
+        this.lifecycle.moveToStarted();
+    }
+
+    void registerNodeConnection(List<TcpChannel> nodeChannels, ConnectionProfile connectionProfile) {
+        TimeValue pingInterval = connectionProfile.getPingInterval();
+        if (pingInterval.millis() < 0) {
+            return;
+        }
+
+        final ScheduledPing scheduledPing = pingIntervals.computeIfAbsent(pingInterval, ScheduledPing::new);
+        scheduledPing.ensureStarted();
+
+        for (TcpChannel channel : nodeChannels) {
+            scheduledPing.addChannel(channel);
+
+            channel.addCloseListener(ActionListener.wrap(() -> {
+                scheduledPing.removeChannel(channel);
+            }));
+        }
+    }
+
+    void receiveKeepAlive(TcpChannel channel) {
+        if (channel.isClient() == false) {
+            sendPing(channel);
+        }
+    }
+
+    long successfulPingCount() {
+        return successfulPings.count();
+    }
+
+    long failedPingCount() {
+        return failedPings.count();
+    }
+
+    private void sendPing(TcpChannel channel) {
+        pingSender.send(channel, pingMessage, new ActionListener<Void>() {
+
+            @Override
+            public void onResponse(Void v) {
+                successfulPings.inc();
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                if (channel.isOpen()) {
+                    logger.debug(() -> new ParameterizedMessage("[{}] failed to send transport ping", channel), e);
+                    failedPings.inc();
+                } else {
+                    logger.trace(() -> new ParameterizedMessage("[{}] failed to send transport ping (channel closed)", channel), e);
+                }
+            }
+        });
+    }
+
+    @Override
+    public void close() {
+        lifecycle.moveToStopped();
+        lifecycle.moveToClosed();
+    }
+
+    private class ScheduledPing extends AbstractLifecycleRunnable {
```

Comment:
can be static no?

Reply:
`ScheduledPing` uses lifecycle, logger, thread pool, etc

Reply:
ok